src/components/name-tag.js

```diff
@@ -0,0 +1,377 @@
+    const matScale = new THREE.Vector3();
+    const mat = new THREE.Matrix4();
+    return function(t) {
+      if (this.model) {
+        this.nametagEl.object3D.getWorldPosition(worldPos);
+        this.neckEl.object3D.matrixWorld.decompose(matPos, matRot, matScale);
+        matScale.set(1, 1, 1);
+        matPos.setY(this.nametagElPosY + this.ikRootEl.object3D.position.y);
+        mat.compose(
+          matPos,
+          matRot,
+          matScale
+        );
+        setMatrixWorld(this.nametagEl.object3D, mat);
+      }
+
+      if (this.nametagVisibility === "showClose") {
+        this.avatarRig.object3D.getWorldPosition(avatarRigWorldPos);
+        this.el.object3D.getWorldPosition(worldPos);
+        this.wasNametagVisible = this.isNametagVisible;
+        this.isNametagVisible = avatarRigWorldPos.sub(worldPos).length() < this.nametagVisibilityDistance;
+        this.updateNameTag();
+      }
+      if (!this.isTalking && this.isTyping) {
+        typingAnimTime = t;
+        this.nametagTypingEl.object3D.traverse(o => {
+          if (o.material) {
+            o.material.opacity = (Math.sin(typingAnimTime / TYPING_ANIM_SPEED) + 1) / 2;
+            typingAnimTime -= TYPING_ANIM_SPEED;
+          }
+        });
+      }
+      if (DEBUG) {
+        this.updateAvatarModelBBAA();
+        this.avatarBBAAHelper.matrixNeedsUpdate = true;
+        this.avatarBBAAHelper.updateMatrixWorld(true);
+      }
+    };
+  })(),
+
+  play() {
+    this.el.addEventListener("model-loading", this.onModelLoading);
+    this.el.addEventListener("model-loaded", this.onModelLoaded);
+    this.el.addEventListener("analyser-volume-updated", this.onAnalyserVolumeUpdated);
+    this.el.sceneEl.addEventListener("presence_updated", this.onPresenceUpdated);
+    window.APP.store.addEventListener("statechanged", this.onStateChanged);
+  },
+
+  pause() {
+    this.el.removeEventListener("model-loading", this.onModelLoading);
+    this.el.removeEventListener("model-loaded", this.onModelLoaded);
+    this.el.removeEventListener("analyser-volume-updated", this.onAnalyserVolumeUpdated);
+    this.el.sceneEl.removeEventListener("presence_updated", this.onPresenceUpdated);
+    window.APP.store.removeEventListener("statechanged", this.onStateChanged);
+  },
+
+  onPresenceUpdated({ detail: presenceMeta }) {
+    if (presenceMeta.sessionId === this.playerSessionId) {
+      this.updateFromPresenceMeta(presenceMeta);
+    }
+  },
+
+  updateFromPresenceMeta(presenceMeta) {
+    this.displayName = presenceMeta.profile.displayName;
+    this.identityName = presenceMeta.profile.identityName;
+    this.isRecording = !!(presenceMeta.streaming || presenceMeta.recording);
+    this.isOwner = !!(presenceMeta.roles && presenceMeta.roles.owner);
+    this.isTyping = !!presenceMeta.typing;
+    this.isHandRaised = !!presenceMeta.hand_raised;
+    this.updateNameTag();
+  },
+
+  updateDisplayName() {
+    if (this.displayName) {
+      this.nametagTextEl.addEventListener(
```

Comment:
Related to the above, since we end up calling this multiple times even when the text doesn't need to be updated, this event listener will end up just sitting around waiting for the event to be fired (`sync()` in Troika does not fire the callback if the text wasn't dirty https://github.com/protectwise/troika/blob/main/packages/troika-three-text/src/Text.js#L402). This would result in a bunch of event listeners hanging around and then all firing at once, when you actually did change your displayname. Its all synchronous so I don't think anything would break but it would do a bunch of unnecessary work.

Reply:
Updated